### ISSUE TITLE
Any part of FLux model can be parallel

### ISSUE THREAD

**Issue by TuanNT-ZenAI** [chengzeyi/ParaAttention#31]
I tested using the same prompt and shape, and the results of Para-attn compared to the base model are very similar in most cases. However, some differences appear, especially when using Flux-schnell with 20 steps.

Do you think the two GPUs might be communicating in other parts of the Flux model, such as RMSNorm, rather than only in the Attention ?

**Comment by chengzeyi** [comment]
@TuanNT-ZenAI This is caused by SplitK and different tile size heristic choices.

**Comment by TuanNT-ZenAI** [comment]
I see in [Distrifusion](https://github.com/mit-han-lab/distrifuser), they not only communicate K and V in Attention, they also communicate in Conv2D and GroupNorm. the results is very close with original image. 
Does Flux can do the same?

**Comment by chengzeyi** [comment]
> I see in [Distrifusion](https://github.com/mit-han-lab/distrifuser), they not only communicate K and V in Attention, they also communicate in Conv2D and GroupNorm. the results is very close with original image. Does Flux can do the same?

FLUX does not have Conv in its transformer blocks and GroupNorm is only channel wise other than token wise.

**Comment by chengzeyi** [comment]
> Yes I mean now in ParaAttn or Piflux, you only communicate in Attention, somewhere we can communicate to achieve results more closely resembling the original image

I think no extra parts are needed.

**Comment by TuanNT-ZenAI** [comment]
Thanks @chengzeyi! I have another question: Why is the final image of Piflux with all steps as sync_step different from ParaAttention? I thought that when communicating the full K and V at every step, the results should be the same.

**Comment by chengzeyi** [comment]
@TuanNT-ZenAI Do you enable Ring Parallelism in ParaAttention? If so, it could cause some precision loss in the reduction behaviour of the output tensor.

**Comment by TuanNT-ZenAI** [comment]
No @chengzeyi , I use 2 GPU and set  
`mesh = init_context_parallel_mesh(
    pipe.device.type,
    # max_ulysses_dim_size=2
)`

the result also has small difference with original image

![Image](https://github.com/user-attachments/assets/8b178fff-db75-4281-9050-e748e8d9a6cc)
![Image](https://github.com/user-attachments/assets/e3167eeb-99c7-4d8d-a83e-a2c7a51b0046)

`import torch
import torch.distributed as dist
from diffusers import FluxPipeline

dist.init_process_group()

torch.cuda.set_device(dist.get_rank())

pipe = FluxPipeline.from_pretrained(
    "black-forest-labs/FLUX.1-schnell",
    torch_dtype=torch.bfloat16,
).to("cuda")

from para_attn.context_parallel import init_context_parallel_mesh
from para_attn.context_parallel.diffusers_adapters import parallelize_pipe
from para_attn.parallel_vae.diffusers_adapters import parallelize_vae

mesh = init_context_parallel_mesh(
    pipe.device.type,
    max_ulysses_dim_size=2
)
parallelize_pipe(
    pipe,
    mesh=mesh,
)

for i in range(5):
    image = pipe(
        "A pair of clear hands looking through a transparent glass Christmas ball , hyper-realistic, minimalist, futuristic background with cute Christmas decorations like Santa Claus, a snowman, and snowflakes, 8k",
        num_inference_steps=8,
        height=1024,
        width=1024,
        generator=torch.Generator(device="cuda").manual_seed(0),
        # output_type="pil" if dist.get_rank() == 0 else "pt",
    ).images[0]

if dist.get_rank() == 0:
    print("Saving image to flux.png")
    image.save("flux_2.png")

dist.destroy_process_group()


# torchrun --nproc_per_node=2 parallel_examples/run_flux.py
`

**Comment by TuanNT-ZenAI** [comment]
I re-test and see piflux full sync step is same with ulysses in para_attn, but all of them is different with original image

**Comment by TuanNT-ZenAI** [comment]
Hi @chengzeyi, I see Piflux with full sync steps is faster than Para-attention when using 4 GPU. Does it make sense? Communication between 4 GPU in Piflux is faster than Para?
